### ISSUE TITLE
Rethink/rewrite rule 942130 with re2 support

### ISSUE THREAD

**Issue by fzipi** [coreruleset/coreruleset#2353]
### Motivation

Allow porters to have more options for using CRS in additional engines.

### Proposed solution

The rule 942130 (SQL Tautology) matches using these regexes:
```
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?=[\s'\"`()]*?(?:\b\1\b)
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?<=>[\s'\"`()]*?(?:\b\1\b)
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?like[\s'\"`()]*?(?:\b\1\b)
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?rlike[\s'\"`()]*?(?:\b\1\b)
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?sounds\s+like[\s'\"`()]*?(?:\b\1\b)
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?regexp[\s'\"`()]*?(?:\b\1\b)
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?!=[\s'\"`()]*?(?!\b\1\b)[\d\w]+
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?<=[\s'\"`()]*?(?!\b\1\b)[\d\w]+
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?>=[\s'\"`()]*?(?!\b\1\b)[\d\w]+
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?<>[\s'\"`()]*?(?!\b\1\b)[\d\w]+
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?<[\s'\"`()]*?(?!\b\1\b)[\d\w]+
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?>[\s'\"`()]*?(?!\b\1\b)[\d\w]+
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?\^[\s'\"`()]*?(?!\b\1\b)[\d\w]+
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?is\s+not[\s'\"`()]*?(?!\b\1\b)[\d\w]+
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?not\s+like[\s'\"`()]*?(?!\b\1\b)[\d\w]+
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?not\s+regexp[\s'\"`()]*?(?!\b\1\b)[\d\w]+
```

- The usage of `\1` is not compatible with re2, but we can probably change it for a named group like this: 
```
[\s'\"`()]*?\b([\d\w]+)\b[\s'\"`()]*?=[\s'\"`()]*?(?:\b\1\b)
```
to 
```
[\s'\"`()]*?\b(?P<one>[\d\w]+)\b[\s'\"`()]*?=[\s'\"`()]*?(?:\b(?P=one)\b)
```
- still we need to think if there is an alternative for the "before text not matching" `(?!\b` that breaks the compatibility

### Additional context

Tests are really simple for this rule in particular, I wonder if there is additional information on how this rule was crafted.

**Comment by fzipi** [comment]
Ok, after going backwards in git, got to commit 0d3eed2 (sqli rulerevision: disassembling sqli rules) by @franbuehler . Can you join me in try to understand what are the constructs `(?!\b\1\b)` trying to match?

**Comment by fzipi** [comment]
I went all over the history of this rule (previously known as 950901), but sadly the meaning is lost in time looks like.

**Comment by fzipi** [comment]
This is the oldest "version" that I could find:
```
 #
 # -=[ SQL Tautologies ]=-
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "(?i:([\s'\"`´’‘\(\)]*   )([\d\w]++)([\s'\"`´’‘\(\)]*)(?:(?:=|<=>|r?like|sounds\s+like|regexp)([\s'\"`´’‘\(\)]*)\2|(?:!=|<=|>=|<>|<|>|\^|is\s+  not|not\s+like|not\s+regexp)([\s'\"`´’‘\(\)]*)(?!\2)([\d\w]+)))" \
         "phase:2,rev:'2',ver:'OWASP_CRS/2.2.6',maturity:'9',accuracy:'8',capture,multiMatch,t:none,t:urlDecodeUni,t:   replaceComments,ctl:auditLogParts=+E,block,msg:'SQL Injection Attack',id:'950901',logdata:'Matched Data: %{TX.0}       found within %{MATCHED_VAR_NAME}: %{MATCHED_VAR}',severity:'2',tag:'OWASP_CRS/WEB_ATTACK/SQL_INJECTION',tag:'WASCTC/   WASC-19',tag:'OWASP_TOP_10/A1',tag:'OWASP_AppSensor/CIE1',tag:'PCI/6.5.2',setvar:'tx.msg=%{rule.msg}',setvar:tx.       sql_injection_score=+%{tx.critical_anomaly_score},setvar:tx.anomaly_score=+%{tx.critical_anomaly_score},setvar:tx.     %{rule.id}-OWASP_CRS/WEB_ATTACK/SQL_INJECTION-%{matched_var_name}=%{tx.0}"
``` 
at commit b9e5ea357178bee58a99dda46352e61222e8ea78 (the very initial import).

**Comment by theseion** [comment]
These expressions match things like `select * from foo where 1 sounds like '1`. The goal is to construct a query that always returns true. Hence, making sure that both "strings" are the same is crucial (the first half uses a "positive" boolean operator (e.g. `=`), the second half a negative one (e.g. `!=`), which is why the second half needs negative look-ahead (`(?!...)`)).

Named back-references aren't supported by re2 either (https://github.com/google/re2/wiki/Syntax).

**Comment by dune73** [comment]
That's good thinking @theseion. Thanks.

**Comment by fzipi** [comment]
I like the macro expansion. Gives us more flexibility. We need to check if this works 🤔

**Comment by fzipi** [comment]
Added a draft PR so we get the feeling on how this looks.

**Comment by fzipi** [comment]
Fixed by #2425
